Sources/System/MachPort.swift

```diff
@@ -0,0 +1,318 @@
+/*
+ This source file is part of the Swift System open source project
+
+ Copyright (c) 2022 Apple Inc. and the Swift System project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+*/
+
+#if $MoveOnly && (os(macOS) || os(iOS) || os(watchOS) || os(tvOS))
+
+import Darwin.Mach
+
+protocol MachPortRight {}
+
+enum Mach {
+    @_moveOnly
+    struct Port<RightType:MachPortRight> {
+        internal var name:mach_port_name_t
+        internal var context:mach_port_context_t
+
+        /// Transfer ownership of an existing unmanaged Mach port right into a
+        /// Mach.Port by name.
+        ///
+        /// This initializer aborts if name is MACH_PORT_NULL.
+        ///
+        /// If the type of the right does not match the type T of Mach.Port<T>
+        /// being constructed, behavior is undefined.
+        ///
+        /// The underlying port right will be automatically deallocated at the
+        /// end of the Mach.Port instance's lifetime.
+        ///
+        /// This initializer makes a syscall to guard the right.
+        init(name:mach_port_name_t) {
+            assert(name != mach_port_name_t(MACH_PORT_NULL))
+            self.name = name
+
+            if (RightType.self == ReceiveRight.self) {
+                let secret = mach_port_context_t(arc4random())
+                let kr = mach_port_guard(mach_task_self_, name, secret, 0)
+                assert(kr == KERN_SUCCESS)
+                self.context = secret
+            }
+            else {
+                self.context = 0
+            }
+        }
+
+        /// Borrow access to the port name in a block that can perform
+        /// non-consuming operations.
+        ///
+        /// Take care when using this function; many operations consume rights,
+        /// and send-once rights are easily consumed.
+        ///
+        /// If the right is consumed, behavior is undefined.
+        ///
+        /// The body block may optionally return something, which will then be
+        /// returned to the caller of withBorrowedName.
+        func withBorrowedName<ReturnType>(body:(mach_port_name_t) -> ReturnType) -> ReturnType {
+            return body(name)
+        }
+
+        deinit {
+            if name != 0xFFFFFFFF /* MACH_PORT_DEAD */ {
+                if RightType.self == ReceiveRight.self {
+                    // recv rights must be mod ref'ed instead of deallocated
+                    let kr = mach_port_unguard(mach_task_self_, name, context)
+                    assert(kr == KERN_SUCCESS)
+
+                    let kr2 = mach_port_mod_refs(mach_task_self_, name, MACH_PORT_RIGHT_RECEIVE, -1)
+                    assert(kr2 == KERN_SUCCESS)
+                } else {
+                    mach_port_deallocate(mach_task_self_, name)
```

Comment:
This branch needs to check the return value?

Reply:
Given how prevalent the mach_call + precondition pairing is in this file, it probably deserves pulling out into a private helper function?

Something like:
```swift
private func machPrecondition(
    _ expected: kern_return_t = KERN_SUCCESS,
    file: StaticString = #file,
    line: Int = #line,
    _ body: @autoclosure () -> kern_return_t
) {
    let kr = body()
    precondition(kr == expected, file: file, line: line)
}
```

Would transform most of these into:
```swift
machPrecondition(mach_port_deallocate(mach_task_self_, name))
```